### ISSUE TITLE
URL isn't sufficient information for useful processing

### ISSUE THREAD

**Issue by tabatkins** [w3c/browser-specs#5]
Bikeshed uses its own bespoke set of specs to look at right now, preserved in a private db maintained by @plinss. It contains a little more information than just URLs, which Bikeshed uses in important ways, and so if Bikeshed and similar tools are meant to consume information from this list, presumably they'd need more info too.

Specifically, the relevant info that's tracked in Bikeshed's current db is:

* /TR url *and* draft/ED URL. (Bikeshed calls these "snapshot" and "current" urls, but I'm not picky about names.)
* shortname, both complete and (unversioned, version #) pair. (These two need to be separate because there's not a consistent pattern for formatting versioned shortnames; "svg2" versus "css-images-3", for example.)
* spec title; currently we track both the "full" title (like "CSS Images & Replaced Content Module Level 3") and a short title (like "CSS Images 3"); I use "full" in the bibliography index but "short" in section autolinks embedded in prose. For many specs these are identical.
* whether the spec is single-page or multi-page. The db tooling goes and discovers the relevant pages if it's a multipage spec; presumably we'd want to just put that info into the db.

**Comment by plinss** [comment]
The spec DB also has the filename of the home page, which is usually Overview.html, but not always. Shepherd's spec parser needs that to understand that internal links to "spec/Overview.html" are actually links to "spec/" and prevents the home page from being parsed twice. I could hard-code 'Overview.html', 'index.html', and whatever else is in the DB at the moment, but I'd rather not.

In addition, the DB has the notion of the "current level" for a spec (which is often not just the last one), and an alternative shortname (which should actually be multiple alternatives), for when specs gets renamed so old URLs and other references can get redirected. Those fields are mostly used by the CSSWG draft server.

**Comment by tidoust** [comment]
Pretty useful to have a vision of what info would be useful to have to make ingestion easy! Boundaries between projects are still fuzzy, I guess part of the whole exercise is about figuring out what these boundaries are and what are the sources of input.

I would separate information that can be automatically derived from information that needs to be manually entered. That is, I would view the `specs.json` file as an initial raw list of specs and would only put things in there that need to be manually maintained. From there, I would document how retrieve additional information about each spec. This repo could both contain tools that can retrieve that information, and the results of processing the initial list to have a list completed with basic information to ease ingestion by other projects (also see https://github.com/w3c/browser-specs/issues/1#issuecomment-576261400).

For instance, as soon as there exists a /TR URL, starting from it, one can infer the shortname and query the W3C API to retrieve the spec title and the URL of the Editor's Draft. The W3C API could be the official source of info there, and I would not duplicate the content in `specs.json` (although, again, this repo could actually export a list of specs completed with that info).

Fetching the spec, it is also relatively easy to determine whether it is a single-page or multi-page spec.

There will always be exceptions to the rule. Some of them will have to be documented and hardcoded in tools (for instance rules to extract levels from shortnames), and in a few cases done on a per-spec basis, meaning that it should definitely be possible to put such information in `specs.json`.

I'm not clear yet about whether Specref should be seen as a source of info on specs listed here, in which case Specref could be used to gather additional information about specs that are not in /TR, or if, in the end, we'd rather want things to be the other way round, meaning info here becoming a source that Specref ingests (noting it already ingests /TR specs). In any case, we should make sure not to duplicate efforts and information.

@tabatkins @plinss, a few questions and comments on the info you need:
- The "short" title case is interesting. Who decides and maintains what the "short" title is? Are we talking about a handful of specs that have a different short title or more than that?
- Out of curiosity, what do you use unversioned shortnames and levels for?
- The notion of levels triggers the question of what to put in the `specs.json` file when multiple levels exist. Some tools will want to see all levels. Other tools will only want to see the latest level (e.g. we want only one IDL extract in Web Platform Tests). When the W3C API allows to easily navigate from unversioned shortnames to versioned ones, listing only unversioned shortname URLs could help with maintenance (and does not prevent exporting a list with all the levels if that seems more useful). In the case of Reffy, it would also be good to know when a spec is a "delta" spec.

**Comment by plinss** [comment]
> I would separate information that can be automatically derived from information that needs to be manually entered. That is, I would view the specs.json file as an initial raw list of specs and would only put things in there that need to be manually maintained. From there, I would document how retrieve additional information about each spec. This repo could both contain tools that can retrieve that information, and the results of processing the initial list to have a list completed with basic information to ease ingestion by other projects (also see #1 (comment)).

What's in Shepherd's spec DB _is_ a list of the information we've needed that we can't reliably derive directly from the specs. There's a ton of additional info there that we do. Believe me, I have no desire to manually maintain this data and wish it could _all_ be automatically derived from the specs. Tab and I have done a lot of work over the years to improve that situation but there is still some metadata that can't be easily pulled from the specs.

> For instance, as soon as there exists a /TR URL, starting from it, one can infer the shortname and query the W3C API to retrieve the spec title and the URL of the Editor's Draft. The W3C API could be the official source of info there, and I would not duplicate the content in specs.json (although, again, this repo could actually export a list of specs completed with that info).

Many EDs exist for some time before there's a version on /TR. Relying on a /TR URL is a non-starter for us.

In general, I agree that I'd much rather have a single source of all this info.

> Fetching the spec, it is also relatively easy to determine whether it is a single-page or multi-page spec.

For many years Shepherd presumed that all specs were multipage. What this means is that links found in the spec that are sub-paths of the spec's main URL are considered other pages of the same spec. A while back we had to add the multipage/single page boolean to prevent certain specs from triggering that behavior. I don't recall the exact situation that triggered this. Maybe Tab does? I'm guessing specs that link to additional docs in the same directory that aren't part of the spec.

We could get rid of this flag if we had a better way of determining which links were part of the same document. Though that would probably require following the link to determine what's there...

> There will always be exceptions to the rule. Some of them will have to be documented and hardcoded in tools (for instance rules to extract levels from shortnames), and in a few cases done on a per-spec basis, meaning that it should definitely be possible to put such information in specs.json.

I really, really, really, don't want to have _any_ exceptions hard-coded in tools. That puts a huge burden on tooling authors and all but guarantees that tools will have implementation differences. Shepherd's spec parser already has a few spec-specific rules for legacy specs, like HTML and SVG, it's a pain.

> @tabatkins @plinss, a few questions and comments on the info you need:

> The "short" title case is interesting. Who decides and maintains what the "short" title is? Are we talking about a handful of specs that have a different short title or more than that?

At this point it's entered by the person adding the spec to the DB. It's generally the colloquial name of the spec and some judgement is applied. I'd much like to see this as metadata embedded in the spec. So far I've found spec authors to not be consistent in following naming conventions tho.

> Out of curiosity, what do you use unversioned shortnames and levels for?

Tab can fill in more of how Bikeshed uses them. The draft server uses them to associate different specs as being the levels of the same spec, it also uses them to auto-generate server aliases.

--

In general, I'm at a point where I have to question the purpose of this list. Currently we have Shepherd's DB, Specref, /TR (and it's XML api), the new W3C API (that I haven't spent much time digging into yet), Biblio.json, and probably something else I'm forgetting. They're all incomplete, and all are missing some information that other tools need. Why make yet another list of incomplete information?

I'd really like to see a single source of truth that contains *all* the info *all* the tools need, without tools needing to keep yet another out-of-sync list of exceptions. If a tool has a use-case for additional info, I want to see a process where the primary list can be expanded rather than see another list pop up somewhere else.

FWIW, if you want to see what's in Shepherd's DB, you can query the API at https://api.csswg.org/shepherd/spec/?draft=1 for just the high-level spec data including drafts. If you want the full dump: https://api.csswg.org/shepherd/spec/?draft=1&anchors=1&links=1&backlinks=1 (currently 64.5MB)

**Comment by tidoust** [comment]
> Many EDs exist for some time before there's a version on /TR. Relying on a /TR URL is a non-starter for us.

Right, what I'm pointing out is that once there is a /TR URL, it becomes a useful starting point. The way we've been dealing with that in Reffy is that we use the ED URL as starting point as long as there is no /TR URL, and then switch to the /TR URL. From time to time, the transition unfortunately breaks a couple of things because the shortname you might infer from an ED URL and use to name things (typically the name of the file that contains the IDL extract and that gets pushed to Web Platform Tests in the case of Reffy) may not be the /TR shortname.


> At this point it's entered by the person adding the spec to the DB. It's generally the colloquial name of the spec and some judgement is applied. I'd much like to see this as metadata embedded in the spec. So far I've found spec authors to not be consistent in following naming conventions tho.

OK, that was the reason why I was asking the question. In an ideal world, this would be metadata in the spec itself.


> In general, I'm at a point where I have to question the purpose of this list. Currently we have Shepherd's DB, Specref, /TR (and it's XML api), the new W3C API (that I haven't spent much time digging into yet), Biblio.json, and probably something else I'm forgetting. They're all incomplete, and all are missing some information that other tools need. Why make yet another list of incomplete information?
>
> I'd really like to see a single source of truth that contains all the info all the tools need, without tools needing to keep yet another out-of-sync list of exceptions.

This list indeed starts as an attempt to create a single source of truth. You have to maintain code and data with exceptions-to-the-rule. We has a similar list of [CSS](https://github.com/tidoust/reffy/blob/master/src/specs/specs-css.json) and [IDL](https://github.com/tidoust/reffy/blob/master/src/specs/specs-idl.json) specs in Reffy that we crawl among other things to create IDL extracts that end up in Web Platform Tests. Other tools do similar things. Each time a new spec appears, it needs to be added to various places.


> If a tool has a use-case for additional info, I want to see a process where the primary list can be expanded rather than see another list pop up somewhere else.

As said earlier, "boundaries between projects are still fuzzy". There is a hell of a lot of information that can be associated with individual specs. For instance, Reffy adds the serialization of the IDL syntax tree when the spec defines IDL terms. Another tool will extract all normative statements and their position in the spec (that seems to generate >50MB of data for the HTML spec alone). Yet another tool might be looking at the history of bugs in the issue tracker of the spec. I agree that these should be seen as expansions of the info associated with specs in the list.

I don't see putting that additional info in this list as an immediate goal. For me, the immediate need is for a list of spec identifiers with core (and up-to-date) info about each spec (the info you both mentioned looks like a good starting point), that could be used as main input list across tools, and not a list that contains all possible knowledge about a spec.

**Comment by tabatkins** [comment]
> I don't see putting that additional info in this list as an immediate goal. For me, the immediate need is for a list of spec identifiers with core (and up-to-date) info about each spec (the info you both mentioned looks like a good starting point), that could be used as main input list across tools, and not a list that contains all possible knowledge about a spec.

Yes, exactly, I'm happy to just have a list of specs, with the relevant spec-based information (not contents). I'd *really* like a canonical list for this that the ecosystem can rely on, because a lot of my hesitation in starting some useful projects in Bikeshed is the difficulty of matching specs across different data sources.

> This repo could both contain tools that can retrieve that information, and the results of processing the initial list to have a list completed with basic information to ease ingestion by other projects 

This sounds ideal, yes. Maintaining *less* data in the "core" list, just what's absolutely necessary to derive the rest, is good, as it's easier to add to and keep from going stale. Just so long as we have the fully-expanded version also available, both for ease-of-use (I'd prefer not to add a node dependency to Bikeshed...) and more importantly, to keep us honest that everything *can* be derived from the core list.

> Out of curiosity, what do you use unversioned shortnames and levels for?

These are *extremely* important for Bikeshed's linking model. When multiple versions of a spec exist, I do *not* want an autolink to a term defined in all of those versions to cause Bikeshed to complain that it can't tell which one you mean; you 99.9% of the time want to link to the latest version. (The logic's a little more complex than that in fact, but that's the core of it.)  (/TR already recognizes this for a lot of specs, as I can navigate to `/TR/css-images/` and get taken to version 4, not 3.) There's a few more bits that similarly care about the latest version of a spec, where earlier versions would confuse/break the feature if they weren't detectable from shortname like this.

> There will always be exceptions to the rule. Some of them will have to be documented and hardcoded in tools (for instance rules to extract levels from shortnames), and in a few cases done on a per-spec basis, meaning that it should definitely be possible to put such information in specs.json.

As @plinss said, if these exceptions trickle down to the *users* of this list, that's unacceptable. But if it's just a matter of "constructing this list can be a little complicated, with some manual intervention required to build", then sure, that's expected.

> I don't recall the exact situation that triggered [multipage detection misfiring]. Maybe Tab does? I'm guessing specs that link to additional docs in the same directory that aren't part of the spec.

I don't remember being involved in anything around this, so I dunno. Your guess sounds valid, tho; "explainers" in the same folder as the spec would trigger that logic.

**Comment by plinss** [comment]
> Right, what I'm pointing out is that once there is a /TR URL, it becomes a useful starting point. The way we've been dealing with that in Reffy is that we use the ED URL as starting point as long as there is no /TR URL, and then switch to the /TR URL. From time to time, the transition unfortunately breaks a couple of things because the shortname you might infer from an ED URL and use to name things (typically the name of the file that contains the IDL extract and that gets pushed to Web Platform Tests in the case of Reffy) may not be the /TR shortname.

Which is a good example of how deriving the shortname from the URL is too fragile and why we track that separately.

> OK, that was the reason why I was asking the question. In an ideal world, this would be metadata in the spec itself.

Agreed.

> As said earlier, "boundaries between projects are still fuzzy". There is a hell of a lot of information that can be associated with individual specs. For instance, Reffy adds the serialization of the IDL syntax tree when the spec defines IDL terms. Another tool will extract all normative statements and their position in the spec (that seems to generate >50MB of data for the HTML spec alone). Yet another tool might be looking at the history of bugs in the issue tracker of the spec. I agree that these should be seen as expansions of the info associated with specs in the list.

> I don't see putting that additional info in this list as an immediate goal. For me, the immediate need is for a list of spec identifiers with core (and up-to-date) info about each spec (the info you both mentioned looks like a good starting point), that could be used as main input list across tools, and not a list that contains all possible knowledge about a spec.

I didn't mean putting all the project specific information into the main list. We really don't want Shepherd's 60MB of anchor and link data there for example. Anyone needs that they can query Shepherd's API. I meant extending the list to add only high-level spec metadata that can't be derived, like the filename of the index page, if it's multipage or not, etc. I expect as new tooling gets created to pull info from specs there will be new needs for metadata that can't be found in the spec, that's what I meant by we need to be able to add fields to the list.

Anything that can be derived directly from the spec, without needing hard-coded exceptions, should be left to the spec itself (and in a stable, and documented manner). And over time, we should start moving that data into the specs as machine parsable metadata, as we've been doing with Shepherd and Bikeshed. e.g. Shepherd has a lot of heuristics to classify definition anchors, much of that logic isn't needed for new specs because Bikeshed generates explicit metadata about definitions.

(As a side note, we should talk about Reffy, seems like it's duplicating a lot of things Shepherd is already doing, like parsing and processing all the IDL in the specs)

**Comment by tidoust** [comment]
I've been looking at the notion of "current" level. Based on the info from Shepherd, I get the following list of specs for which the "current" level is not the "last" level:

* `css-animations-2` - [CSS Animations 2](https://drafts.csswg.org/css-animations-2/) => [CSS Animations 1](https://www.w3.org/TR/css3-animations/)
* `css-backgrounds-4` - [CSS Backgrounds 4](https://drafts.csswg.org/css-backgrounds-4/) => [CSS Backgrounds 3](https://www.w3.org/TR/css3-background/)
* `css-break-4` - [CSS Fragmentation 4](https://www.w3.org/TR/css-break-4/) => [CSS Fragmentation 3](https://www.w3.org/TR/css-break-3/)
* `css-color-5` - [CSS Color 5](https://www.w3.org/TR/css-color-5/) => [CSS Color 4](https://www.w3.org/TR/css-color-4/)
* `css-conditional-4` - [CSS Conditional 4](https://www.w3.org/TR/css-conditional-4/) => [CSS Conditional 3](https://www.w3.org/TR/css3-conditional/)
* `css-gcpm-4` - [CSS GCPM 4](https://drafts.csswg.org/css-gcpm-4/) => [CSS GCPM 3](https://www.w3.org/TR/css-gcpm-3/)
* `css-grid-2` - [CSS Grid Layout 2](https://www.w3.org/TR/css-grid-2/) => [CSS Grid Layout 1](https://www.w3.org/TR/css-grid-1/)
* `css-images-4` - [CSS Images 4](https://www.w3.org/TR/css4-images/) => [CSS Images 3](https://www.w3.org/TR/css3-images/)
* `css-multicol-2` - [CSS Multicol 2](https://drafts.csswg.org/css-multicol-2/) => [CSS Multicol 1](https://www.w3.org/TR/css3-multicol/)
* `css-overflow-4` - [CSS Overflow 4](https://www.w3.org/TR/css-overflow-4/) => [CSS Overflow 3](https://www.w3.org/TR/css-overflow-3/)
* `css-page-4` - [CSS Paged Media 4](https://drafts.csswg.org/css-page-4/) => [CSS Paged Media 3](https://www.w3.org/TR/css3-page/)
* `css-shapes-2` - [CSS Shapes 2](https://drafts.csswg.org/css-shapes-2/) => [CSS Shapes 1](https://www.w3.org/TR/css-shapes-1/)
* `css-sizing-4` - [CSS Sizing 4](https://drafts.csswg.org/css-sizing-4/) => [CSS Sizing 3](https://www.w3.org/TR/css-sizing-3/)
* `css-text-4` - [CSS Text 4](https://drafts.csswg.org/css-text-4/) => [CSS Text 3](https://www.w3.org/TR/css-text-3/)
* `css-text-decor-4` - [CSS Text Decoration 4](https://drafts.csswg.org/css-text-decor-4/) => [CSS Text Decoration 3](https://www.w3.org/TR/css-text-decor-3/)
* `css-transforms-2` - [CSS Transforms 2](https://www.w3.org/TR/css-transforms-2/) => [CSS Transforms 1](https://www.w3.org/TR/css-transforms-1/)
* `css-transitions-2` - [CSS Transitions 2](https://drafts.csswg.org/css-transitions-2/) => [CSS Transitions 1](https://www.w3.org/TR/css3-transitions/)
* `css-typed-om-2` - [CSS Typed OM 2](https://drafts.css-houdini.org/css-typed-om-2/) => [CSS Typed OM 1](https://www.w3.org/TR/css-typed-om-1/)
* `filter-effects-2` - [Filter Effects 2](https://drafts.fxtf.org/filter-effects-2/) => [Filter Effects 1](https://www.w3.org/TR/filter-effects-1/)
* `mediaqueries-5` - [Media Queries 5](https://www.w3.org/TR/mediaqueries-5/) => [Media Queries 4](https://www.w3.org/TR/mediaqueries-4/)

In all (*) of these cases, I believe that the last level is a "delta" spec. Having the info about whether a spec is a delta spec seems more meaningful to me than tracking the current level of a spec. That is, I know that the lists of CSS/IDL/terms are, by definition, not yet complete in a delta spec, whereas I'm not sure what it means for the last level of a spec not to be the current level. It certainly means that it's not ready but... not ready for what and who?

I prepared PR #8 to allow specs to be defined as "delta" in `specs.json`. The PR also adds a `previousLevel` property to link to the previous level. For instance, the data returned for Filter Effects Level 2 (provided that spec gets flagged as a delta one) would become:

```json
  {
    "url": "https://drafts.fxtf.org/filter-effects-2/",
    "name": "filter-effects-2",
    "shortname": "filter-effects",
    "level": 2,
    "delta": true,
    "previousLevel": "filter-effects-1"
  }
```

In other words, when the `delta` flag is set, the `previousLevel` property would point at the "current" level. @tabatkins @plinss, does that seem good or did I miss something?

(*) There is actually one apparent exception to the rule: [CSS Fragmentation 4](https://www.w3.org/TR/css-break-4/) does not look like a delta spec over [CSS Fragmentation 3](https://www.w3.org/TR/css-break-3/). Or at least, it no longer seems to be a delta spec. Is there a specific reason that explains why level 4 is not yet the current level of the CSS Fragmentation module?

**Comment by plinss** [comment]
The notion of whether a spec is a delta spec or not is useful. But it’s somewhat orthogonal to whether the last version of a spec is considered current or not. There is usually, but not always, a high correlation between the two.

We make delta specs when someone is working on new features, and one (or more) previous spec level is still undergoing some amount of churn. It’s simply a mechanism to prevent the editor of a future level from having to track changes in a previous version that isn’t all that stable.

The decision of which level is considered “current work” is a judgement call made by the WG. Generally when “current work” shifts to the next level, it’s a signal that the previous level is stable and it’s no longer necessary to have a delta spec, but not always.

Conflating the two concepts in tooling will impose a work mode on authors that they may not otherwise choose, e.g. you have to stop being a delta spec in order to be considered “current work”. I don’t think it’s proper for tooling to impose this restriction just to save a db field.

Also, FWIW, I wouldn’t store the “delta” state as a Boolean, make it an enum, even if there are only two states currently, experience shows there will likely be more in the future.

**Comment by tabatkins** [comment]
Yeah, "current" just means it's the version we think is appropriate to show when you visit the unversioned url.

**Comment by tidoust** [comment]
@plinss wrote:
> The notion of whether a spec is a delta spec or not is useful. But it’s somewhat orthogonal to whether the last version of a spec is considered current or not. There is usually, but not always, a high correlation between the two.

@tabatkins wrote:
> Yeah, "current" just means it's the version we think is appropriate to show when you visit the unversioned url.

Understood. So, "current" is the level targeted by the [canonical URL](https://www.w3.org/2005/05/tr-versions#shortnames) for /TR/ specs, although with a slightly broadened definition here (I think the canonical URL can only redirect to /TR/ documents whereas we might want to redirect to a level that has not yet been published to /TR/).

I'll expose some additional `currentLevel` property which, by default, will target the "latest non delta" spec, with a mechanism to force the current level manually in the initial list of specs.

In the list I pasted, I note that the unversioned /TR/ URL does not redirect to the appropriate "current" level for:
- [CSS Image](https://www.w3.org/TR/css-images/) - redirects to level 4, not to level 3
- [CSS Overflow](https://www.w3.org/TR/css-overflow/) - redirects to level 4, not to level 3
- [CSS Text Decoration](https://www.w3.org/TR/css-text-decor/) - redirects to level 4, not to level 3


> Also, FWIW, I wouldn’t store the “delta” state as a Boolean, make it an enum, even if there are only two states currently, experience shows there will likely be more in the future.

Hmm, I'm trying to balance things here. I like having a Boolean because it simplifies writing conditional logic in most languages. What could these additional states look like in the future?

I could perhaps imagine that we flag delta specs that are "pure delta" (meaning that they don't have any content defined in previous level) differently from delta specs that are "overlapping delta" (meaning that they include some of the contents defined in previous level), but then "pure delta" seems too strict (by definition, there will be some minimal amount of content copied over from the previous level in most delta specs).

Proliferation of states does not seem a good thing in general, because it makes the situation confusing for external readers. People have been, and are still, struggling with the existence of TR and ED versions. Levels and delta specs introduce two more dimensions. In other words, I actually hope that groups won't introduce more states in the future.

**Comment by plinss** [comment]
> Understood. So, "current" is the level targeted by the canonical URL for /TR/ specs, although with a slightly broadened definition here (I think the canonical URL can only redirect to /TR/ documents whereas we might want to redirect to a level that has not yet been published to /TR/).

I believe it's more used for draft specs. The CSSWG has a custom built draft server with CI that does those redirects/aliases. I expect as we get better about pushing FPWDs we'll get into the situation on /TR as well.

I don't thing it'll be common (or desired at all?) for delta specs to wind up on /TR.

> I'll expose some additional currentLevel property which, by default, will target the "latest non delta" spec, with a mechanism to force the current level manually in the initial list of specs.

Sounds good.

> Hmm, I'm trying to balance things here. I like having a Boolean because it simplifies writing conditional logic in most languages. What could these additional states look like in the future?

It's not clear that we'll have additional states, I'm speaking more from experience of having defined boolean states and getting bitten when a new state comes up. Booleans may be convenient, but they're fragile and become very inconvenient when new states are needed. FWIW, CSS does not have booleans for this reason. Just because we can't define the additional states now doesn't mean we won't come up with one as editing practices evolve.

I also don't think `relationToPrevLevel` is the right concept, in my head it's more of a `whatKindOfSpecIsThis` (maybe `composition`?) maybe states could be something like: 'delta', 'evolution' (or 'extension'), 'replacement' (for specs that completely replace rather than extend a previous level). We don't need to define all of them now, that's the point.

FWIW Shepherd also has a field: `work_status` enum('completed', 'stable', 'testing', 'refining', 'revising', 'exploring', 'rewriting', 'abandoned').
